src/PJ_pipeline.c

```diff
@@ -477,14 +477,35 @@ PJ *OPERATION(pipeline,0) {
         }
     }
 
-    /* Check that output units from step i are compatible with expected units in step i+1 */
+    /* Check that output units from step i are compatible with expected units in step i-1 and i+1 */
     for (i = 1; i < nsteps; i++) {
-        enum pj_io_units unit_returned = pj_right (P->opaque->pipeline[i]);
-        enum pj_io_units unit_expected = pj_left  (P->opaque->pipeline[i+1]);
+        enum pj_io_units previous_step_input  = i>1 ? pj_left  (P->opaque->pipeline[i-1]) : PJ_IO_UNITS_WHATEVER;
+        enum pj_io_units previous_step_output = i>1 ? pj_right (P->opaque->pipeline[i-1]) : PJ_IO_UNITS_WHATEVER;
+
+        enum pj_io_units current_step_input  = pj_left  (P->opaque->pipeline[i]);
+        enum pj_io_units current_step_output = pj_right (P->opaque->pipeline[i]);
+
+        enum pj_io_units next_step_input  = pj_left  (P->opaque->pipeline[i+1]);
```

Comment:
should i+1 < nsteps being tested ?

Reply:
Are you referring to the if/else statements below?

Reply:
no, it seems to me that P->opaque->pipeline[i+1] could be an out-of-bound access when i == steps - 1